From 36f7f6b35ae4d5c7c61baa0f93f16d398c1a23d5 Mon Sep 17 00:00:00 2001
From: Ivan Kruglov <mail@ikruglov.com>
Date: Tue, 17 Jun 2025 06:23:18 -0700
Subject: [PATCH] cgroup-util: Always open cgroupv2 attribute files in
 O_NONBLOCK mode

backport of https://github.com/systemd/systemd/pull/37211

As explained in https://lore.kernel.org/all/20250419183545.1982187-1-shakeel.butt@linux.dev/, writing to memory.max or memory.high triggers synchronous memory reclaim if the limit is lowered. This can end up taking nonnegligible amounts of time, completely blocking pid1 from doing any other work while the reclaim is ongoing.

To address this problem, the kernel going to add O_NONBLOCK semantics to memory.max and memory.high. If the file is opened with O_NONBLOCK, the synchronous memory reclaim is skipped and only triggered later without blocking the process writing the file. Let's make sure we make use of this by opening cgroupv2 attribute files with O_NONBLOCK.

We opt to do this for all cgroupv2 attribute files, to make sure that if the same problem happens elsewhere in the future and is fixed in the same way, we immediately take advantage of that fix without having to make changes in systemd as well. We probably never want to block when writing cgroupv2 attributes and any cases where we do want to block should indicate so explicitly instead of blocking by default.
---
 src/basic/cgroup-util.c | 7 ++++++-
 src/basic/fileio.c      | 3 ++-
 src/basic/fileio.h      | 1 +
 3 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/basic/cgroup-util.c b/src/basic/cgroup-util.c
index 309dccb45ab22..e79beedeb222d 100644
--- a/src/basic/cgroup-util.c
+++ b/src/basic/cgroup-util.c
@@ -1721,7 +1721,12 @@ int cg_set_attribute(const char *controller, const char *path, const char *attri
         if (r < 0)
                 return r;
 
-        return write_string_file(p, value, WRITE_STRING_FILE_DISABLE_BUFFER);
+        /* https://lore.kernel.org/all/20250419183545.1982187-1-shakeel.butt@linux.dev/ adds O_NONBLOCK
+         * semantics to memory.max and memory.high to skip synchronous memory reclaim when O_NONBLOCK is
+         * enabled. Let's always open cgroupv2 attribute files in nonblocking mode to immediately take
+         * advantage of this and any other asynchronous resource reclaim that's added to the cgroupv2 API in
+         * the future. */
+        return write_string_file(p, value, WRITE_STRING_FILE_DISABLE_BUFFER|WRITE_STRING_FILE_OPEN_NONBLOCKING);
 }
 
 int cg_get_attribute(const char *controller, const char *path, const char *attribute, char **ret) {
diff --git a/src/basic/fileio.c b/src/basic/fileio.c
index 2d29c384cc7a6..8beb55963a381 100644
--- a/src/basic/fileio.c
+++ b/src/basic/fileio.c
@@ -335,7 +335,8 @@ int write_string_file_full(
                         (FLAGS_SET(flags, WRITE_STRING_FILE_NOFOLLOW) ? O_NOFOLLOW : 0) |
                         (FLAGS_SET(flags, WRITE_STRING_FILE_CREATE) ? O_CREAT : 0) |
                         (FLAGS_SET(flags, WRITE_STRING_FILE_TRUNCATE) ? O_TRUNC : 0) |
-                        (FLAGS_SET(flags, WRITE_STRING_FILE_SUPPRESS_REDUNDANT_VIRTUAL) ? O_RDWR : O_WRONLY),
+                        (FLAGS_SET(flags, WRITE_STRING_FILE_SUPPRESS_REDUNDANT_VIRTUAL) ? O_RDWR : O_WRONLY) |
+                        (FLAGS_SET(flags, WRITE_STRING_FILE_OPEN_NONBLOCKING) ? O_NONBLOCK : 0),
                         mode,
                         &made_file);
         if (fd < 0) {
diff --git a/src/basic/fileio.h b/src/basic/fileio.h
index 71ec6ba9210a0..31c90dc531485 100644
--- a/src/basic/fileio.h
+++ b/src/basic/fileio.h
@@ -29,6 +29,7 @@ typedef enum {
         WRITE_STRING_FILE_MODE_0444                  = 1 << 11,
         WRITE_STRING_FILE_SUPPRESS_REDUNDANT_VIRTUAL = 1 << 12,
         WRITE_STRING_FILE_LABEL                      = 1 << 13,
+        WRITE_STRING_FILE_OPEN_NONBLOCKING           = 1 << 14,
 } WriteStringFileFlags;
 
 typedef enum {
